Issue #104 — Support persistence of `MMEEvent` for `telemetryMetrics` events

alfwatt opened the pull request:
PR for: https://github.com/mapbox/mapbox-events-ios/issues/104

Commit:
Issue #105 - Refactor `MMENSDateWrapper` into `MMEDate`

Commit:
Implement the required `NSDate` subclass overrides and protocols
Add `offsetFromServer` on `MMEDate` to track the offset at the time a date is created
Remove wrapper instances and get test running

Commit:
Tests for MMEDate, remove all DateWrapper references

Commit:
MMEDate - better description and coding keys
MMEEvent - include the date the event was created, some documentation
MMEMetrics - move time slip tracking to MMEDate

rclee committed:
Update MapboxMobileEventsTests/MMEEventsManagerTests.mm

Co-Authored-By: alfwatt <alf@istumbler.net>

Commit:
Review: fix nits in .h, use constants for encode/decode, check coded version

Commit:
Consolidate Date formatters into MMEDate
MMEMetrics.date -> MMEMetrics.recordingStarted

Commit:
Constant key values and version numbers for Coding/Decoding

rclee commented on MapboxMobileEvents/MMEDate.h:
```diff
@@ -0,0 +1,49 @@
+/*! @brief returns the recorded time offset from the server's time frame */
++ (NSTimeInterval)recordedTimeOffsetFromServer;
+
+
```
nit: whitespace police

alfwatt replied:
Which whitespace where?

rclee replied:
I think this was resolved when rebasing. :magic:

rclee commented on MapboxMobileEvents/MMEMetricsManager.m:
```diff
@@ -106,25 +104,17 @@ - (void)updateCoordinate:(CLLocationCoordinate2D)coordinate {
 
 - (void)resetMetrics {
-    self.metrics = [[MMEMetrics alloc] init];
+    self.metrics = [MMEMetrics new];
```
Probably just being nitpicky here -- `init` is used everywhere else in the repo, are we moving to `new` instead?

alfwatt replied:
I like using new in place of the `[[id alloc] init]` pair, it's been in Foundation forever

rclee replied:
That's fine with me if that's the preference. We do use some custom initializers (which I don't think is possible with `new`?) in some places but we just handle them on a case by case basis. https://github.com/mapbox/mapbox-events-ios/blob/master/MapboxMobileEvents/MMEConfigurator.h#L14

rclee commented on MapboxMobileEventsTests/MMEDateTests.mm:
```diff
@@ -0,0 +1,74 @@
+    context(@"+ recordTimeOffsetFromServer:", ^{
+        it(@"computes offsets from server time", ^{
+            NSDate* serverTime = [NSDate dateWithTimeIntervalSinceNow:interval];
+            NSTimeInterval recorded = [MMEDate recordTimeOffsetFromServer:serverTime];
```
nit: Some of this common setup could go into the above `beforeEach` (`NSDate* serverTime = [NSDate dateWithTimeIntervalSinceNow:interval];` at least).

rclee commented on MapboxMobileEventsTests/MMEDateTests.mm:
```diff
@@ -0,0 +1,74 @@
+    });
+
+    context(@"- mme_oneDayLater", ^{
+        it(@"", ^{
```
Missing a description (maybe "should set the date for the beginning of tomorrow"?) and `context` description name changed above `- mme_oneDayLater` -> `- mme_startOfTomorrow`.

rclee commented on MapboxMobileEventsTests/MMEDateTests.mm:
```diff
@@ -0,0 +1,74 @@
+            NSTimeInterval recorded = [MMEDate recordedTimeOffsetFromServer];
+
+            round(computed) should equal(round(interval));
+            round(recorded) should equal(round(interval));
```
Should we separate these asserts into a second `it` block?

rclee commented on MapboxMobileEvents/MMEMetricsManager.m:
```diff
@@ -141,30 +131,31 @@ - (NSDictionary *)attributes {
+    [combinedAttributes setObject:[MMEDate.iso8601DateFormatter stringFromDate:NSDate.date] forKey:@"created"];
     MMEEvent *debugEvent = [MMEEvent debugEventWithAttributes:attributes];
-    [[MMEEventLogger sharedLogger] logEvent:debugEvent];
+    [MMEEventLogger.sharedLogger logEvent:debugEvent];
```
👍

rclee commented on MapboxMobileEvents/MMEDate.m:
```diff
@@ -0,0 +1,166 @@
+#import "MMEDate.h"
+
+static NSTimeInterval _timeOffsetFromServer = 0.0; // TODO maintain a list of MMEDates with offsets
```
Is this TODO for another PR?

rclee reviewed (commented):
Few small things.

Commit:
Don't re-declare NSSecureCoding on MMEDate

Commit:
Don't redundantly declare NSCoding, nits in decode

Commit:
Sort Source Files by name…
Remove principal class entry from plist

Commit:
Overriding NSCoder on Foundation objects is not without peril
NSDate is a cluster and overrides `-[NSObject classForCoder]`, MMEDate now over-overrides it
Updated tests for secure coding, passing

Commit:
#if blocks around some UIKit calls

Commit:
Add 	`MMEEventsTests.mm`, bring `MMEEventsConfigurationTests.mm` in from the cold
Update tests to reflect new concept of event equality, don't auto-populate the date field in events to fix other tests
Indent inline dictionaries for easier reading

Commit:
Improve MMEDateTests
Add NSSecureCoding test to MMEEventTests

Commit:
Add `-MMEEventManager sendPendingTelemetryMetricsEvent`
Add `-MMEMetricsManager loadPendingTelemetryMetricsEvent`
Write a pending telemetry metrics event when calling `generateTelemetryMetricsEvent`

Commit:
Send any pending telemetry events when resuming metrics collection
If blocks which don't immediately follow an open brace get an blank line

rclee committed:
telemetryMetrics improvements (#113)

* Reformat failedRequests metric 

This should allow the failedRequests metric to be more queryable.

* Track received data 

Implement the event schema changes to begin tracking received data on all requests. 

Tested using breakpoints to track each `data.length` from start to finish.

* Fix bug in event counting

Although this wouldn't be much an issue in our current implementation; once we add in deferred telemetry we would have missed counting any additional batches in APIClient.

Also adds a failed event count if the request is ever nil.

* Implement suggested changes 

Also adds missing tests and removed the synchronous NSURLConnection call that was being used for testing. 

Changed transferred bytes to sent bytes -- this will be reflected in the schema soon.

* Move telemetryMetrics send request method

Moves `telemetryMetrics` send request method to `flush` to have the event send more regularly. 

Previously this method was moved out of `flush` to be more readable but we missed a vital part that `flush` is also called on a timer and `sendTelemetryMetricsEvent` was out of that loop.

* Add telemetryMetrics timer test

Adds a unit test to ensure that we attempt to send a telemetryMetrics event when the event queue timer fires.

Commit:
Generate telemetry events every time we update the telemetry object, write to a file if it's not time to send the data back yet

Commit:
Don't trigger `-Wformat` in the `mapbox-gl-native` build

Commit:
Metrics Manager doesn't know enough about when to generate telemetry events, but the API client does

Commit:
Send pending events as soon as the API Client is initialized

Commit:
Check to see if the file was written

Commit:
Use the a secure archiver explicitly

Commit:
Move the event record into `Library/Caches`
Create the directory where the pending event is saved

Commit:
Securely decode to the correct format

Commit:
Don't reset the metrics when sending a pending event

Commit:
Rebase misses

julianrex commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -1,10 +1,23 @@
+}
+
++ (instancetype)eventWithName:(NSString *)eventName attributes:(NSDictionary *)attributes {
+    return [MMEEvent.alloc initWithDate:MMEDate.date name:eventName attributes:attributes];
```
Why not call `eventWithDate:name:attributes:` here?

julianrex commented on MapboxMobileEvents/MMEEventsManager.m:
```diff
@@ -306,43 +327,70 @@ - (void)sendTurnstileEvent {
+
+    if (pendingMetricsEvent) {
+        __weak __typeof__(self) weakSelf = self;
+        [self.apiClient postEvent:pendingMetricsEvent completionHandler:^(NSError * _Nullable error) {
```
~Using `self` here.~ EDIT: Same mistake here, apologies.

julianrex commented on MapboxMobileEvents/MMEEventsManager.m:
```diff
@@ -306,43 +327,70 @@ - (void)sendTurnstileEvent {
     [MMEEventLogger.sharedLogger logEvent:turnstileEvent];
     
     __weak __typeof__(self) weakSelf = self;
     [self.apiClient postEvent:turnstileEvent completionHandler:^(NSError * _Nullable error) {
```
~Using `self` here.~ EDIT: Whoops - got lost.

Commit:
Review changes

Commit:
Fix tests

rclee commented on MapboxMobileEvents/MMEMetricsManager.m:
```diff
@@ -168,12 +233,12 @@ - (NSDictionary *)attributes {
         attributes[MMEEventDeviceLon] = @(self.metrics.deviceLon);
     }
-    
+    attributes[MMEEventKeyVendorID] = [MMEEventsManager sharedManager].commonEventData.vendorId;
```
This should be removed. I added when I thought it was a different metric -- it's not needed and not accepted by our schema. (equates to `userId`)

alfwatt replied:
Must have picked it up in a rebase conflict, removing

rclee commented on MapboxMobileEvents/MMEMetricsManager.m:
```diff
@@ -182,30 +247,75 @@ - (NSDictionary *)attributes {
-    if (zeroHour.timeIntervalSinceNow > 0) {
-        NSString *debugDescription = [NSString stringWithFormat:@"TelemetryMetrics event isn't ready to be sent; waiting until %@ to send", zeroHour];
+    NSString *metricsDate = [MMEDate.iso8601DateFormatter stringFromDate:NSDate.date];
+    MMEEvent *telemetryMetrics = [MMEEvent telemetryMetricsEventWithDateString:metricsDate attributes:self.attributes];
```
If we'll be generating the telemetry event more often then we need to add the same logic for the zeroHour to eventsManager `sendTelemetryMetricsEvent:` or else the event will sent more often than 24 hour chunks.

rclee replied:
`sendTelemetryMetricsEvent` logic was faulty in this way -- should have had the conditional there in the first place I think.

alfwatt replied:
`generateTelemetryMetricsEvents` doesn't return the event if it's not time, it just writes the pending event and returns nil so that `sendTelemetryMetrics` doesn't need to know about scheduling

rclee replied:
Ahh, you're right -- sorry about that. Thanks for the walkthrough.

rclee reviewed (changes requested):
Couple things here.

rclee reviewed (dismissed):
LGTM 👍

Commit:
Guard writing pending events on older iOS versions
Add preflight checks to the Makefile
